channeldb/payment_control.go

```diff
@@ -290,18 +290,19 @@ func (p *PaymentControl) RegisterAttempt(paymentHash lntypes.Hash,
 			return err
 		}
 
-		// Ensure the payment is in-flight.
-		if err := ensureInFlight(p); err != nil {
-			return err
-		}
-
 		// We cannot register a new attempt if the payment already has
-		// reached a terminal condition:
+		// reached a terminal condition. We check this before
+		// ensureInFlight because it is a more general check.
```

Comment:
From [this line](https://github.com/lightningnetwork/lnd/blob/master/channeldb/payments.go#L353), suppose the payment has two HTLC attempts, `h1`, `h2`, then,

`h1` settled, and

- `h2` settled, then we have, `inflight=false && settled=true`, go to `case !inflight&& settled` => `StatusSucceeded`
- `h2` failed => then we have, `inflight=false && settle=true`, go to `case !inflight && settled` => `StatusSuccedded`
- `h2` is neither settled or failed, then we have, `inflight=true && settle=true`, go to `default` => `StatusInFlight`

`h1` failed, and

- `h2` settled, then we have `inflight=false && settled=true`, go to `case !inflight && settled` => `StatusSucceeded`
- `h2` failed, then we have `inflight=false && settled=false`, then,
  - if `failureReason != nil` , => `StatusFailed`
  - otherwise, `StatusInFlight`
- `h2` is neither settled or failed, then we have `inflight=true && settled=false`, => `StatusInFlight`

`h1` is neither settled or failed

- `h2` settled, then we have `inflight=true && settled=true`, go to `default` => `StatusInFlight`
- `h2` failed, then we have `inflight=true && settled=false`, go to `default` => `StatusInFlight`
- `h2` is neither settled or failed, then we have `inflight=true && settled=false`, go to `default` => `StatusInFlight`

Reply:
My question is, why are we relying on the payment's `failureReason`but not the `HTLCFailInfo.Failre` here?

Reply:
Yep, you got it right 👌 

> why are we relying on the payment's failureReason but not the HTLCFailInfo.Failre here?

Not entirely sure what you mean, but the `failureReason` holds the overall payment failure state, while a HTLC failure doesn't really impact the payment failure state (since we will just retry with new HTLCs). The only exception is when an HTLC failure indicates that we should terminate the payment, in which case we must wait for no HTLCs to be in flight before we can determine the final state of the payment.